charlotte_core/src/acpi/tables.rs

```diff
@@ -0,0 +1,96 @@
+//! ACPI tables handling
+
+use core::{mem, str};
+
+use super::{rsdp, AcpiTables};
+
+#[repr(C, packed)]
+#[derive(Copy, Clone)]
+pub struct SDTHeader {
+    signature: [u8; 4],
+    length: u32,
+    revision: u8,
+    checksum: u8,
+    oem_id: [u8; 6],
+    oem_table_id: [u8; 8],
+    oem_revision: u32,
+    creator_id: u32,
+    creator_revision: u32,
+}
+
+impl SDTHeader {
+    pub fn new(tables: &mut AcpiTables, sig: [u8; 4]) -> Option<Self> {
```

Comment:
The entire algorithm for this doesn't really make a whole lot of sense. 

It would probably instead be better to create a struct for each table and store them in the `AcpiTables` struct. For example the XSDT/RSDT struct could look like:

```
pub struct Xsdt {
    header: SDTHeader,
    n_entries: usize,
    addr_width: usize, // Needed because the RSDP contains 32 bit addresses, while the XSDT contains 64 bit addresses
}
```

Then after parsing the RSBP, we parse the X/RSDT into this struct, and store it in `AcpiTables`. Then for any other table like the FADT for it's `new` function we just search through the XSDT entries for it. The XSDT in memory is essentially just the header and then a bunch of pointers to the other tables, which we just search until we find the table with the correct signature. Each SDT follows a similar structure of header followed by data (which is structured differently for each table).